### PR TITLE
Enhanced Session Recording

### PR DESCRIPTION
**Pull request by russjones**
**Description**

See the [Enhanced Session Recording](https://docs.google.com/document/d/1JhS5_AiRYUm0mH5rveB7i5tvP_MJZ6iA_JlzM7tvCxA) design document for full details.

### CODE REVIEW

File: lib/bpf/bpf.go [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+		events.PID:        event.PID,
+		events.CgroupID:   event.CgroupID,
+		events.Program:    command,
+		events.Path:       path,
+		events.Flags:      event.Flags,
+		events.ReturnCode: event.ReturnCode,
+	}
+	ctx.AuditLog.EmitAuditEvent(events.SessionDisk, eventFields)
+
+	//// Remove, only for debugging.
```
Comment (klizhentas) [inline lib/bpf/bpf.go]: extra code

File: lib/events/sessionlog.go [hunk 2]
```diff
@@ -320,57 +420,136 @@ func EventFromChunk(sessionID string, chunk *SessionChunk) (EventFields, error)
...
+	// Update index for the last event that was processed.
+	sl.enhancedIndexes[chunk.EventType] = chunk.EventIndex
+
+	// Convert to a marshallable event.
+	fields, err := EventFromChunk(sessionID, chunk)
 	if err != nil {
 		return -1, trace.Wrap(err)
 	}
-	_, err = sl.eventsFile.Write(append(bytes, '\n'))
+	bytes, err = json.Marshal(fields)
```
Comment (klizhentas) [inline lib/events/sessionlog.go]: fast marshal here?

File: lib/bpf/common.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	ServerID string
+
+	// Login is the Unix login for this session.
+	Login string
+
+	// User is the Teleport user.
+	User string
+
+	// PID is the process ID of Teleport when it re-executes itself. This is
+	// used by Telepor to find itself by cgroup.
```
Comment (cove) [inline lib/bpf/common.go]: @russjones "Telepor"

File: lib/bpf/disk.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+struct val_t {
+    u64 pid;
+    char comm[TASK_COMM_LEN];
+    const char *fname;
+    int flags;
+};
+
+struct data_t {
+    u64 cgroup;
+    u64 pid;
```
Comment (cove) [inline lib/bpf/disk.go]: pid_t is an int

File: lib/bpf/bpf.go [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+			events.SessionEventID:  ctx.SessionID,
+			events.SessionServerID: ctx.ServerID,
+			events.EventLogin:      ctx.Login,
+			events.EventUser:       ctx.User,
+			// Exec fields.
+			events.PID:        event.PPID,
+			events.PPID:       event.PID,
+			events.CgroupID:   event.CgroupID,
+			events.Program:    command,
+			events.Path:       argv[0],
```
Comment (alex-kovoy) [inline lib/bpf/bpf.go]: Does `events.Path` contain the command path or will it be empty (where path is part of the `program`) ?

File: lib/bpf/network.go [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = attachRetProbe(e.module, "tcp_v6_connect", "trace_connect_v6_return")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Open perf buffer and start processing IPv4 events.
+	e.v4EventCh, e.v4LostCh, err = openPerfBuffer(e.module, e.perfMaps, pageCount, "ipv4_events")
```
Comment (alex-kovoy) [inline lib/bpf/network.go]: Just curious if you considered having 1 buffer for all event types? Where you could have 2 tables:  one table for event payloads (byte stream), and another for event definitions containing an event type, payload size with an offset into the payload table. This way we can have a central message pump loop which would dispatch events for further processing. This part would be completely separated.

I think it would make it easier to add new events in the future where you would emit a new event in the bpf program and whoever processes the events may/may-not choose to handle it.

Basically, we can have:
1 way to define all event types
1 way to emit all types of events
1 way to pump all event types and dispatch them (where we can read in batches and dispatch to multiple handlers)

File: lib/bpf/disk.go (hunk 4 above)
Comment (fspmarshall) [reply lib/bpf/disk.go]: The [`bpf_get_current_pid_tgid`](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#4-bpf_get_current_pid_tgid) call seems to return `pid`/`tgid` as packed `u32`s.  I can't find the *rational* documented anywhere, but all the bpf code, even the kernel source, seems to follow the paradigm.

**Comment by fspmarshall** [comment]
Build currently fails on linux if `bcc` is not present.  In particular:

- `lib/bpf/helper.go` is scoped to `linux`, pretty sure it should be scoped to `bpf,linux`.
- `lib/bpf/bpf_nop.go` is scoped to `!linux`, pretty sure it should be scoped to `!bpf`.
- `lib/bpf/bpf_nop.go` imports and does not use `lib/events`.

Fixing the above issues makes the build work, but a bunch of tests still fail.  Unsure about the root issue, but given the errors, I'd guess that some of the tests don't play nicely with `bpf_nop.go`.

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by russjones** [comment]
@gravitational-jenkins retest this please
